fairlearn/postprocessing/_threshold_optimizer.py

```diff
@@ -427,6 +261,178 @@ def _threshold_optimization_equalized_odds(sensitive_features, labels, scores, g
     return predicted_EO_by_sensitive_feature
 
 
+_SUPPORTED_MITIGATIONS = {
+    DEMOGRAPHIC_PARITY: _threshold_optimization_demographic_parity,
+    EQUALIZED_ODDS: _threshold_optimization_equalized_odds}
+
+
+def _get_soft_predictions(estimator, X):
+    """Returns soft predictions of a classifier using either `predict_proba`
+    or `decision_function` methods.
+    """
+    if hasattr(estimator, "predict_proba"):
+        return estimator.predict_proba(X)
+    elif hasattr(estimator, "decision_function"):
+        return estimator.decision_function(X)
+    else:
+        raise ValueError("{} provides neither a `predict_proba` nor a "
+                            "`decision_function` method.".format(
+                                estimator.__class__.__name__
+                            ))
+
+
+class ThresholdOptimizer(ClassifierMixin, BaseEstimator):
+    """A classifier based on the threshold optimization approach.
+
+    The procedure followed is described in detail in
+    `Hardt et al. (2016) <https://arxiv.org/abs/1610.02413>`_.
+
+    :param unconstrained_predictor: The trained predictor whose output will be
+    postprocessed
+    :type unconstrained_predictor: A trained predictor
+
+    :param estimator: An untrained estimator that will be trained, and
+        subsequently its output will be postprocessed
+    :type estimator: An untrained estimator, default=SGDClassifier(loss='log')
+
+    :param grid_size: The number of ticks on the grid over which we evaluate the
+        curves. A large grid_size means that we approximate the actual curve, so
+        it increases the chance of being very close to the actual best solution.
+    :type grid_size: int
+
+    :param flip: Allow flipping to negative weights if it improves accuracy.
+    :type flip: bool
+
+    :param plot: Show ROC/selection-error plot if True
+    :type plot: bool
+
+    :param warm_start: Avoid refitting the underlying estimator if it's already
+        done in a previous fit.
+    :type warm_start: bool, default=False
+
+    :param random_state: set to a constant for reproducibility
+    :type random_state: int, np.RandomState, default=None
+    """
+
+    def __init__(self, *, estimator=SGDClassifier(loss='log'),
+                 constraints=DEMOGRAPHIC_PARITY,
```

Comment:
If we're using SGDClassifier as a default, shall we run it with `random_state=0` (or something of the sort) to make it replicate? Alternatively, I don't think we necessarily need to have a default here.

Reply:
I'm not very set on the `SGDClassifier`, really no hard feelings there. We could instead use `HistGradientBoostingClassifier` which is a much more modern one and people seem to like it.

Regarding the random state, since the user can set the `estimator` parameter, they can set it while passing a random state to it. It's the same in GridSearch and the underlying estimator for example. If the user needs a deterministic result, they should pass an rng to both.

Reply:
Should we have a default at all? Is this commonly done?

Reply:
I'm in favor of dropping the default value for the estimator. It might even make sense to drop a default on constraints.

Reply:
sklearn's common tests assume all estimators can be constructed using default values, except for the `estimator` or a meta-estimator. So we could leave this as required, and follow the [private] protocol to make the common tests understand how it should construct it. It'll use a `Ridge` if the meta-estimator is a regressor and a `LinearDiscriminantAnalysis` if a classifier.

Alternatively, the tests on the class can be disabled/skipped and only test instances.

Reply:
I think that the right approach is to go the meta-estimator route; this is something we will also want to use in our reductions, so if we implement it here, we can propagate it there.

On the API level, I think that `ThresholdOptimizer` should be similar to `CalibratedClassifierCV` [link](https://github.com/scikit-learn/scikit-learn/blob/b194674c42d54b26137a456c510c5fdba1ba23e0/sklearn/calibration.py#L32)

Re. defaults, in `CalibratedClassifierCV`, the `base_estimator` is defaulted to `None`, which becomes `LinearSVC(random_state=0)` during `fit`. I'm happy to go with whatever plays better with sklearn; if we need to have some default, I think that any linear model would work (LinearSVC, LinearDiscriminantAnalysis, or SGDClassifier), I just thought it would be nicer if the default was deterministic.